hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/command/InsertIntoHoodieTableCommand.scala

```diff
@@ -209,19 +209,32 @@ object InsertIntoHoodieTableCommand {
       .getOrElse(INSERT_DROP_DUPS_OPT_KEY.defaultValue)
       .toBoolean
 
-    val operation = if (isOverwrite) {
-      if (table.partitionColumnNames.nonEmpty) {
-        INSERT_OVERWRITE_OPERATION_OPT_VAL  // overwrite partition
-      } else {
-        INSERT_OPERATION_OPT_VAL
+    val enableBulkInsert = parameters.getOrElse(DataSourceWriteOptions.SQL_ENABLE_BULK_INSERT.key,
+      DataSourceWriteOptions.SQL_ENABLE_BULK_INSERT.defaultValue()).toBoolean
+    val isPartitionedTable = table.partitionColumnNames.nonEmpty
+    val isPrimaryKeyTable = primaryColumns.nonEmpty
+    val operation =
+      (isPrimaryKeyTable, enableBulkInsert, isOverwrite, dropDuplicate) match {
+        case (true, true, _, _) =>
+          throw new IllegalArgumentException(s"Table with primaryKey can not use bulk insert.")
+        case (_, true, true, _) if isPartitionedTable =>
+          throw new IllegalArgumentException(s"Insert Overwrite Partition can not use bulk insert.")
+        case (_, true, _, true) =>
+          throw new IllegalArgumentException(s"Bulk insert cannot support drop duplication." +
+            s" Please disable $INSERT_DROP_DUPS_OPT_KEY and try again.")
+        // if enableBulkInsert is true, use bulk insert for the insert overwrite non-partitioned table.
+        case (_, true, true, _) if !isPartitionedTable => BULK_INSERT_OPERATION_OPT_VAL
+        // insert overwrite partition
+        case (_, _, true, _) if isPartitionedTable => INSERT_OVERWRITE_OPERATION_OPT_VAL
+        // insert overwrite table
+        case (_, _, true, _) if !isPartitionedTable => INSERT_OVERWRITE_TABLE_OPERATION_OPT_VAL
+        // if the table has primaryKey and the dropDuplicate has disable, use the upsert operation
+        case (true, false, false, false) => UPSERT_OPERATION_OPT_VAL
+        // if enableBulkInsert is true and the table is non-primaryKeyed, use the bulk insert operation
+        case (false, true, _, _) => BULK_INSERT_OPERATION_OPT_VAL
+        // for the rest case, use the insert operation
+        case (_, _, _, _) => INSERT_OPERATION_OPT_VAL
```

Comment:
Here is my thought on choosing the right operation. Having too many case statements might complicate things and is error prone too. As I mentioned earlier, we should try to do any valid conversions in HoodiesSparkSqlWriter. Only those thats applicable just to sql dml, we should keep it here. 
Anyways, here is one simplified approach. Ignoring the primary, non primary key table for now. We can come back to that later once we have consensus on this. 

We need just two configs. 
hoodie.sql.enable.bulk_insert (default false)
hoodie.sql.overwrite.entire.table (default true)

From sql syntax, there are two commands allowed. 
"INSERT" into and "INSERT OVERWRITE".

"INSERT" with no other configs set -> insert operation
"INSERT" with enable bulk insert set -> bulk_insert
"INSERT OVERWRITE" with no other configs set -> insert_overwrite_table operation
"INSERT OVERWRITE" with hoodie.sql.overwrite.entire.table = false -> insert_overwrite operation.
"INSERT OVERWRITE" with enable bulk_insert set -> bulk_insert. pass the right save mode to HoodieSparkSqlWriter
"INSERT OVERWRITE" with enable bulk_insert set and hoodie.sql.overwrite.entire.table = false -> bulk_insert. pass the right save mode to HoodieSparkSqlWriter.

Reply:
@vinothchandar : do check this out before reviewing other feedbacks.

Reply:
having configs to control sql behavior sounds pretty complex to me. Can we just pick the right operation for each type.

Reply:
actually I came across [INSERT OVERWRITE DIRECTORY](https://spark.apache.org/docs/latest/sql-ref-syntax-dml-insert-overwrite-directory.html) which can be mapped to insert_overwrite. 

Here is a suggestion w/o using any additional configs: 
CTAS -> bulk_insert 
Insert into -> insert
INSERT OVERWRITE -> insert overwrite table
INSERT OVERWRITE DIRECTORY -> insert overwrite

Reply:
I did go through every case here and have 2 suggestions. rest of the cases looks good. You don't need to consider my proposal above. But would like you to consider below feedback. 
1.
```
case (true, true, _, _) if !isNonStrictMode => throw new IllegalArgumentException(s"Table with primaryKey can not use bulk insert in strict mode.")
```
Can we enable preCombine here and proceed with Bulk_Insert operation. Within hudi, we can do preCombine/dedup. As we agreed on using bulk_insert as default with CTAS, this will be a very common use-case. 

2. 
```
case (_, true, true, _) if isPartitionedTable =>
          throw new IllegalArgumentException(s"Insert Overwrite Partition can not use bulk insert.")
```
since we agreed on enabling Bulk_insert as default for CTAS, this will be very common use-case as well. Can you help me understand why do we fail this call? why can't we let it proceed. This is basically, CTAS for a partitioned table.